File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
  ans = [0 for _ in range(len(prices))]
  stack = []
  for i in range(len(prices)) :
    while len(stack) > 0 and prices[stack[-1]] > prices[i] :
      cur_i = stack.pop()
      ans[cur_i] = i - cur_i
      
    stack.append(i)
    
  for si in stack :
    ans[si] = len(prices) - 1 - si
    
  return ans

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int n = prices.length;
        int[] answer = new int[n];
        ArrayDeque<Integer> stack = new ArrayDeque<Integer>();
        for(int i = 0; i < n; i++) {
            while(!stack.isEmpty() && prices[stack.peek()] > prices[i]) {
                int cur = stack.pop();
                answer[cur] = i - cur;
            }
            stack.push(i);
        }
        while(!stack.isEmpty()) {
            int cur = stack.pop();
            answer[cur] = n - cur - 1;
		}
        
        return answer;
    }
}